packages/embark/src/lib/modules/ipfs/index.js

```diff
@@ -28,10 +33,32 @@ class IPFS {
     this.registerUploadCommand();
     this.listenToCommands();
     this.registerConsoleCommands();
-    this.startProcess((err, newProcessStarted) => {
-      this.addStorageProviderToEmbarkJS();
-      this.addObjectToConsole();
-      this.events.emit("ipfs:process:started", err, newProcessStarted);
+    this.events.request("processes:register", "ipfs", {
+      launchFn: (cb) => {
+        if(this.usingRunningNode) {
+          return cb(__("IPFS process is running in a separate process and cannot be started by Embark."));
+        }
+        this.startProcess((err, newProcessStarted) => {
+          this.addStorageProviderToEmbarkJS();
```

Comment:
This function call and the next one cause side-effects asynchronously but they're invoked in a fire-and-forget manner. To avoid race conditions should they happen one after the other before emitting the event on L44?

Reply:
Nice catch! IIRC, this change was made consciously (in a PR a little while back, but can't seem to find the exact one) as the callback for the process manager can be fired once the process itself has been started. The functions `this.addStorageProviderToEmbarkJS();` and `this.addObjectToConsole();` do asynchronous side-effects, but they are not pertinent to process manager. If we needed to wait to callback once the entire module was initialised, then we would definitely need to use a callback chain before firing the callback.